mp/src/gameui/OptionsSubAudio.cpp

```diff
@@ -81,234 +83,249 @@ COptionsSubAudio::~COptionsSubAudio()
+            qb->SetOKButtonText("#GameUI_ChangeLanguageRestart_OkButton");
+            qb->SetCancelButtonText("#GameUI_ChangeLanguageRestart_CancelButton");
+            qb->AddActionSignalTarget(GetParent()->GetParent()->GetParent());
+            qb->DoModal();
+        }
+    }
+}
 
-   if ( nUpdatedAudioLanguage != m_nCurrentAudioLanguage )
-   {
-      // Store new language in static member so that it can be accessed during shutdown when this instance is gone
-      m_pchUpdatedAudioLanguage = (char *) GetLanguageShortName( nUpdatedAudioLanguage );
-      
-      // Inform user that they need to restart in order change language at this time
-      QueryBox *qb = new QueryBox( "#GameUI_ChangeLanguageRestart_Title", "#GameUI_ChangeLanguageRestart_Info", GetParent()->GetParent()->GetParent() );
-      if (qb != NULL)
-      {
-         qb->SetOKCommand( new KeyValues( "Command", "command", "RestartWithNewLanguage" ) );
-         qb->SetOKButtonText( "#GameUI_ChangeLanguageRestart_OkButton" );
-         qb->SetCancelButtonText( "#GameUI_ChangeLanguageRestart_CancelButton" );
-         qb->AddActionSignalTarget( GetParent()->GetParent()->GetParent() );
-         qb->DoModal();
-      }
-   }
+void COptionsSubAudio::ApplySoundQuality()
+{
+    switch (m_pSoundQualityCombo->GetActiveItemUserData()->GetInt("quality"))
+    {
+    case SOUNDQUALITY_LOW:
+        m_cvarDSPSlowCPU.SetValue(true);
+        m_cvarSndPitchQuality.SetValue(false);
+        break;
+    case SOUNDQUALITY_MEDIUM:
+        m_cvarDSPSlowCPU.SetValue(false);
+        m_cvarSndPitchQuality.SetValue(false);
+        break;
+    default:
+        Assert("Undefined sound quality setting.");
+    case SOUNDQUALITY_HIGH:
+        m_cvarDSPSlowCPU.SetValue(false);
+        m_cvarSndPitchQuality.SetValue(true);
+        break;
+    };
+
+    ApplyEnhanceStereo();
 }
 
-//-----------------------------------------------------------------------------
-// Purpose: Called on controls changing, enables the Apply button
-//-----------------------------------------------------------------------------
-void COptionsSubAudio::OnControlModified()
+void COptionsSubAudio::ApplySpeakerSetup()
 {
-	PostActionSignal(new KeyValues("ApplyButtonEnable"));
+    m_cvarSndSurroundSpeakers.SetValue(m_pSpeakerSetupCombo->GetActiveItemUserData()->GetInt("speakers"));
+    ApplyEnhanceStereo();
 }
 
-//-----------------------------------------------------------------------------
-// Purpose: returns true if the engine needs to be restarted
-//-----------------------------------------------------------------------------
-bool COptionsSubAudio::RequiresRestart()
+void COptionsSubAudio::ApplyEnhanceStereo()
 {
-	// nothing in audio requires a restart like now
-	return false;
+    int speakers = m_pSpeakerSetupCombo->GetActiveItemUserData()->GetInt("speakers");
+    int quality = m_pSoundQualityCombo->GetActiveItemUserData()->GetInt("quality");
```

Comment:
Could the `GetActiveItemUserData` ever return null?

Reply:
You bet. Oddly enough the function it directly returns the result of, `GetItemUserData`, is null checked elsewhere in the code. Will add null checks for these.